README.markdown
Introduction
------------

_pegdown_ is a pure Java library for clean and lightweight [Markdown] processing based on a [parboiled] PEG parser.

_pegdown_ is nearly 100% compatible with the original Markdown specification and fully passes the original Markdown test suite.
On top of the standard Markdown feature set _pegdown_ implements a number of extensions similar to what other popular
Markdown processors offer. You can also extend _pegdown_ by your own plugins!
Currently _pegdown_ supports the following extensions over standard Markdown:

* SMARTS: Beautifies apostrophes, ellipses ("..." and ". . .") and dashes ("--" and "---")
* QUOTES: Beautifies single quotes, double quotes and double angle quotes (&laquo; and &raquo;)
* SMARTYPANTS: Convenience extension enabling both, SMARTS and QUOTES, at once.
* ABBREVIATIONS: Abbreviations in the way of [PHP Markdown Extra](http://michelf.com/projects/php-markdown/extra/#abbr).
* ANCHORLINKS: Generate anchor links for headers by taking the first range of alphanumerics and spaces.
* HARDWRAPS: Alternative handling of newlines, see [Github-flavoured-Markdown]
* AUTOLINKS: Plain (undelimited) autolinks the way [Github-flavoured-Markdown] implements them.
* TABLES: Tables similar to [MultiMarkdown](http://fletcherpenney.net/multimarkdown/) (which is in turn like the [PHP Markdown Extra](http://michelf.com/projects/php-markdown/extra/#table) tables, but with colspan support).
* DEFINITION LISTS: Definition lists in the way of [PHP Markdown Extra](http://michelf.com/projects/php-markdown/extra/#def-list).
* FENCED CODE BLOCKS: Fenced Code Blocks in the way of [PHP Markdown Extra](http://michelf.com/projects/php-markdown/extra/#fenced-code-blocks) or [Github-flavoured-Markdown].
* HTML BLOCK SUPPRESSION: Suppresses the output of HTML blocks.
* INLINE HTML SUPPRESSION: Suppresses the output of inline HTML elements.
* WIKILINKS: Support `[[Wiki-style links]]` with a customizable URL rendering logic.
* STRIKETHROUGH: Support ~~strikethroughs~~ as supported in Pandoc and Github.
* ATXHEADERSPACE: Require a space between the `#` and the header title text, as per [Github-flavoured-Markdown]. Frees up `#` without a space to be just plain text.
* FORCELISTITEMPARA: Wrap a list item or definition term in `<p>` tags if it contains more than a simple paragraph.
* RELAXEDHRULES: allow horizontal rules without a blank line following them.
* TASKLISTITEMS: parses bullet lists of the form `* [ ]` and `* [x]` to create GitHub like task list items:
    * [ ] open task item
    * [x] closed or completed task item.
    * [X] also closed or completed task item.
* EXTANCHORLINKS: Generate anchor links for headers using complete contents of the header.
    * Spaces and non-alphanumerics replaced by `-`, multiple dashes trimmed to one.
    * Anchor link is added as first element inside the header with empty content: `<h1><a name="header"></a>header</h1>`
                        
Note: _pegdown_ differs from the original Markdown in that it ignores in-word emphasis as in

    > my_cool_file.txt
    > 2*3*4=5

Currently this "extension" cannot be switched off.


Installation
------------

You have two options:

* Download the JAR for the latest version from [here](http://repo1.maven.org/maven2/org/pegdown/pegdown/).
  _pegdown_ 1.6.0 has only one dependency: [parboiled for Java][parboiled], version 1.1.7.
   
* The pegdown artifact is also available from maven central with group id **org.pegdown** and artifact-id **pegdown**.


Usage
-----

Using _pegdown_ is very simple: Just create a new instance of a [PegDownProcessor] and call one of its
`markdownToHtml` methods to convert the given Markdown source to an HTML string. If you'd like to customize the
rendering of HTML links (Auto-Links, Explicit-Links, Mail-Links, Reference-Links and/or Wiki-Links), e.g. for adding
`rel="nofollow"` attributes based on some logic you can supply your own instance of a [LinkRenderer] with the call
to `markdownToHtml`.

You can also use pegdown only for the actual parsing of the Markdown source and do the serialization to the
target format (e.g. XML) yourself. To do this just call the `parseMarkdown` method of the [PegDownProcessor] to obtain
the root node of the Abstract Syntax Tree for the document.
With a custom [Visitor] implementation you can do whatever serialization you want. As an example you might want to
take a look at the [sources of the ToHtmlSerializer][ToHtmlSerializer].

Note that the first time you create a [PegDownProcessor] it can take up to a few hundred milliseconds to prepare the
underlying parboiled parser instance. However, once the first processor has been built all further instantiations will
be fast. Also, you can reuse an existing [PegDownProcessor] instance as often as you want, as long as you prevent
concurrent accesses, since neither the [PegDownProcessor] nor the underlying parser is thread-safe.

See <http://sirthias.github.com/pegdown/api> for the pegdown API documentation.


Plugins
-------

Since parsing and serialisation are two different things there are two different plugin mechanisms, one for the parser,
and one for the [ToHtmlSerializer]. Most plugins would probably implement both, but it is possible that a plugin might
just implement the parser plugin interface.

For the parser there are two plugin points, one for inline plugins (inside a paragraph) and one for block plugins. These
are provided to the parser using the [PegDownPlugins] class. For convenience of use this comes with its own builder.
You can either pass individual rules to this builder (which is what you probably would do if you were using Scala
rules), but you can also pass it a parboiled Java parser class which implements either [InlinePluginParser] or
[BlockPluginParser] or both. [PegDownPlugins] will enhance this parser for you, so as a user of a plugin you just need
to pass the class to it (and the arguments for that classes constructor, if any). To implement the plugin, you would
write a normal parboiled parser, and implement the appropriate parser plugin interface. You can extend the pegdown
parser, this is useful if you want to reuse any of its rules.

For the serializer there is [ToHtmlSerializerPlugin] interface. It is called when a node that the [ToHtmlSerializer]
doesn't know how to process is encountered (i.e. one produced by a parser plugin). Its `accept` method is passed the
node, the visitor (so if the node contains child nodes they can be rendered using the parent) and the printer for the
plugin to print to. The `accept` method returns true if it knew how to handle the node or false if otherwise and
the [ToHtmlSerializer] loops through each plugin breaking when it reaches one that returns true and if it finds none
throws an exception like it used to.

As an very simple example you might want to take a look at the [sources of the PluginParser test class][PluginParser].


Parsing Timeouts
----------------

Since Markdown has no official grammar and contains a number of ambiguities the parsing of Markdown source, especially
with enabled language extensions, can be "hard" and result, in certain corner cases, in exponential parsing time.
In order to provide a somewhat predictable behavior *pegdown* therefore supports the specification of a parsing timeout,
which you can supply to the [PegDownProcessor] constructor.

If the parser happens to run longer than the specified timeout period it terminates itself with an exception, which
causes the `markdownToHtml` method to return `null`. Your application should then deal with this case accordingly and,
for example, inform the user.

The default timeout, if not explicitly specified, is 2 seconds.


IDE Support
-----------

The excellent [idea-markdown plugin] for IntelliJ IDEA, RubyMine, PhpStorm, WebStorm, PyCharm and appCode uses _pegdown_
as its underlying parsing engine. The plugin gives you proper syntax-highlighting for markdown source and shows you
exactly, how _pegdown_ will parse your texts.


Credits
-------

A large part of the underlying PEG grammar was developed by John MacFarlane and made available with his
tool [peg-markdown](http://github.com/jgm/peg-markdown).   


License
-------

_pegdown_ is licensed under [Apache License 2.0](http://www.apache.org/licenses/LICENSE-2.0).


Patch Policy
------------

Feedback and contributions to the project, no matter what kind, are always very welcome.
However, patches can only be accepted from their original author.
Along with any patches, please state that the patch is your original work and that you license the work to the pegdown
project under the project’s open source license.
  
   [Markdown]: http://daringfireball.net/projects/markdown/ "Main Markdown site"
   [parboiled]: http://www.parboiled.org
   [Github-flavoured-Markdown]: http://github.github.com/github-flavored-markdown/
   [MultiMarkdown]: http://fletcherpenney.net/multimarkdown/users_guide/multimarkdown_syntax_guide/
   [PegDownProcessor]: http://www.decodified.com/pegdown/api/org/pegdown/PegDownProcessor.html
   [LinkRenderer]: http://www.decodified.com/pegdown/api/org/pegdown/LinkRenderer.html
   [Visitor]: http://www.decodified.com/pegdown/api/org/pegdown/ast/Visitor.html
   [ToHtmlSerializer]: https://github.com/sirthias/pegdown/blob/master/src/main/java/org/pegdown/ToHtmlSerializer.java
   [idea-markdown plugin]: https://github.com/nicoulaj/idea-markdown
   [SBT]: http://www.scala-sbt.org/
   [Node]: http://www.decodified.com/pegdown/api/org/pegdown/ast/Node.html
   [PegDownPlugins]: http://github.com/sirthias/pegdown/blob/master/src/main/java/org/pegdown/plugins/PegDownPlugins.java
   [InlinePluginParser]: http://github.com/sirthias/pegdown/blob/master/src/main/java/org/pegdown/plugins/InlinePluginParser.java
   [BlockPluginParser]: http://github.com/sirthias/pegdown/blob/master/src/main/java/org/pegdown/plugins/BlockPluginParser.java
   [ToHtmlSerializerPlugin]: http://github.com/sirthias/pegdown/blob/master/src/main/java/org/pegdown/plugins/ToHtmlSerializerPlugin.java
   [PluginParser]: http://github.com/sirthias/pegdown/blob/master/src/test/java/org/pegdown/PluginParser.java


CHANGELOG
Version 1.6.0 (2015-09-18)
--------------------------
- Fixed collision between ANCHORLINKS and STRIKETHROUGH, WIKILINKS extensions (#161)
- Fixed handling of image `alt` attributes
- Improved parsing and rendering of nested lists in numerous ways
- Improved HTML tag recognition
- Fixed problem in Definition List parsing (#180)
- Fixed escaping of special chars used by disabled extensions (#179)
- Extended `StrikeNode` with `isClosed` member (#178)
- Fixed reference node including trailing blank lines (#177)
- Fixed Setext header handling to allow trailing spaces/tabs (#176)
- Fixed `CaptionNode` not having source index range set (#175)
- Added `ATXHEADERSPACE` extension to require a space/tab after # in AtxHeaders (#144)
- Added `RELAXEDHRULES` extenstion to relax requirement of a blank line after horiz. rule. (#129)
- Fixed `STRIKETHROUGH` extension not adding `~` to special characters (#131)
- Fixex Wiki Link alternate format (#182)
- Fixed `TableCellNode` not including trailing `|` (#183)
- Fixed block quotes not followed by another block quote including trailing blank lines (#184)
- Added `TASKLISTITEMS` optional extension to handle GFM style task list items (#185)
- Added `isSetext` to `HeaderNode` to distiguish between Atx and Setext types (#186)
- Added `ALL_OPTIONALS` and `ALL_WITH_OPTIONALS` extension set
- Added `EXTANCHORS` optional extension
- Fixed two or more blank lines between list items breaking the list into separate lists (#190)
- Fixed paragraph wrapping for list items would produce an AST different from how blank lines before list items (#192)
- Changed `private` to `protected` methods in `ToHtmlSerializer` so that it can be extended without duplicating code (#193)


Version 1.5.0 (2013-12-0)
-------------------------
- Upgraded to parboiled 1.1.7
- Fixed problem with recognizing Wikilinks
- Fixed problem in image tag HTML rendering
- Added anchor links to headers
- Added timeout check in label rule (#104)
- Added new helper method to PegDownProcessor to supply list of ToHtmlSerializerPlugin's to ToHtmlSerializer.


Version 1.4.2 (2013-12-09)
--------------------------
- Fix autolinks not ending at `<` character (#102)
- Enable LinkRenderer to influence image rendering (#95)
- Add STRIKETHROUGH extension
- Upgrade to parboiled 1.1.6


Version 1.4.1 (2013-07-20)
--------------------------
- Added support for extending SpecialChar rule via plugin


Version 1.4.0 (2013-06-23)
--------------------------
- Added basic plugin support (see #77, thx to James Roper and also Jakub Jirutka)


Version 1.3.0 (2013-06-04)
--------------------------
- Upgraded to parboiled 1.1.5
- Improved handling of emph/strong (fixes #43, #65 and #78, thx a lot to Elmer van Chastelet for the patch!)
- Added extensible, type-specific serialization for code blocks (thx to Jared Bunting)


Version 1.2.1 (2012-11-27)
--------------------------
- Upgraded to parboiled 1.1.4
- Added support for caption on tables (thx to Geoffrey Picron)


Version 1.2.0 (2012-10-05)
--------------------------
- Upgraded to parboiled 1.1.3
- Implemented parsing timeout, closes #42
- Fixed IndexOutOfBoundsException in ToHtmlSerializer, closes #39
- Changed parser to not swallow backslash before non-special characters, fixes #35
- Added test for fenced code blocks containing an empty line, closes #61
- Migrated all tests from TestNG to specs2
- Switched build tool from Apache Buildr to SBT


Version 1.1.0 (2011-10-05)
--------------------------
- Upgraded to parboiled 1.0.2
- Introduced LinkRenderer layer for easy customization of HTML link rendering
- Added support for [[Wiki-style]] links, thx to Brett Porter
- Fixed #27 (Missing emphasis when paragraph has multiple lines)
- Fixed #28 (Parser not fully extensible)
- Fixed #32, #33 (HTML tags with attribute names containing underscores are not recognized as tags, thx to Edward Samson)
- Fixed #34 (Pathologically slow parsing when strong and italics emphasis are applied together)


Version 1.0.2 (2011-07-05)
--------------------------
- Simplify custom extensions by making `private` fields/methods of ToHtmlSerializer and Parser `protected`
- Fixed #20 (support GFM fenced code blocks)
- Fixed #21 (simplify integration of custom code transformers)
- Fixed #22 (some characters/sequences break links when AutoLinks extensions is on)
- Fixed broken Java 1.5 compatibility (thx to Julien Nicoulaud)
- Fixed maven build: add testng and jtidy deps for tests, and configure the compiler plugin for java 5 (thx to Lukas Theussl)


Version 1.0.1 (2011-05-17)
--------------------------
- Fixed #18 (code blocks in list items split by lines)


Version 1.0.0 (2011-05-13)
--------------------------
- Closed #3 (PhpMarkdownExtra-style definition lists)
- closed #14 (superfluous paragraph breaks)
- Closed #15 (email obfuscation)
- Closed #16 (PhpMarkdownExtra-style fenced code blocks)
- Closed #17 ('nofollow' links)
- Added buffer indices to AST nodes, enabling markdown syntax highlighting
- Removed tabstop expansion (prevented proper AST node index building)
- Improved single quoting rules for Smartypants extension
- Completely refactored list parsing for code size, performance and simplicity
- Fixed small problem with AutoLink delimiters
- Upgraded to parboiled 1.0.0


Version 0.9.2 (2011-04-06)
--------------------------
- Fix: special characters not encoded correctly in plain text sections
- Fix: HTML hex entities incorrectly parsed
- Add indices into underlying buffer to AST nodes
- Upgraded to parboiled 0.11.1


Version 0.9.1 (2011-03-21)
--------------------------
- Upgraded to parboiled 0.11.0


Version 0.9.0 (2011-02-11)
--------------------------
- Completely overhauled and cleaned up general architecture, split out HTML serialization into AST visitor (thanks to Ravindar Reddy)
- Fixed bug causing the disappearance of text elements in certain list constructs
- Fixed excessive backtracking bug on parsing larger blocks of 'opening-tags-only' HTML

src/org/pegdown/AstNodeType.java
/*
 * Copyright (C) 2010 Mathias Doenitz
 *
 * Based on peg-markdown (C) 2008-2010 John MacFarlane
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.pegdown;

/**
 * Convenience interface defining the AST node type constants as well as a lookup array for reverse naming resolution. 
 */
public interface AstNodeType {

    static int DEFAULT = 0;

    // basic elements
    static int APOSTROPHE = 1;
    static int ELLIPSIS = 2;
    static int EMDASH = 3;
    static int ENDASH = 4;
    static int HTML = 5;
    static int LINEBREAK = 6;
    static int SPACE = 7;
    static int SPECIAL = 8;
    static int TEXT = 9;

    // inline groups
    static int CODE = 10;
    static int EMPH = 11;
    static int H1 = 12;
    static int H2 = 13;
    static int H3 = 14;
    static int H4 = 15;
    static int H5 = 16;
    static int H6 = 17;
    static int STRONG = 18;
    static int SINGLE_QUOTED = 19;
    static int DOUBLE_QUOTED = 20;

    // blocks
    static int BLOCKQUOTE = 21;
    static int HRULE = 22;
    static int HTMLBLOCK = 23;
    static int PARA = 24;
    static int VERBATIM = 25;
    static int ABBREVIATION = 26;

    // lists
    static int BULLET_LIST = 27;
    static int ORDERED_LIST = 28;
    static int TIGHT_LIST_ITEM = 29;
    static int LOOSE_LIST_ITEM = 30;

    // links
    static int AUTO_LINK = 31;
    static int EXP_LINK = 32;
    static int EXP_IMG_LINK = 33;
    static int MAIL_LINK = 34;
    static int REF_LINK = 35;
    static int REF_IMG_LINK = 36;

    static int LINK_LABEL = 37;
    static int LINK_REF = 38;
    static int LINK_TITLE = 39;
    static int LINK_URL = 40;
    static int REFERENCE = 41;

    // reverse resolution
    static String[] TYPE_NAMES = new String[]{
            "DEFAULT",
            "APOSTROPHE", "ELLIPSIS", "EMDASH", "ENDASH", "HTML", "LINEBREAK", "SPACE", "SPECIAL", "TEXT",
            "CODE", "EMPH", "H1", "H2", "H3", "H4", "H5", "H6", "STRONG", "SINGLE_QUOTED", "DOUBLE_QUOTED",
            "BLOCKQUOTE", "HRULE", "HTMLBLOCK", "PARA", "VERBATIM", "ABBREVIATION",
            "BULLET_LIST", "ORDERED_LIST", "TIGHT_LIST_ITEM", "LOOSE_LIST_ITEM",
            "AUTO_LINK", "EXP_LINK", "EXP_IMG_LINK", "MAIL_LINK", "REF_LINK", "REF_IMG_LINK",
            "LINK_LABEL", "LINK_REF", "LINK_TITLE", "LINK_URL", "REFERENCE"
    };

}
